storage/watchable_store.go

```diff
@@ -448,13 +383,105 @@ type watcher struct {
 	ch chan<- WatchResponse
 }
 
-// unsafeAddWatcher puts watcher with key k into watchableStore's synced.
-// Make sure to this is thread-safe using mutex before and after.
-func unsafeAddWatcher(synced *map[string]map[*watcher]struct{}, k string, wa *watcher) error {
+// watchKVs raw key bytes and storagepb.KeyValue bytes
+// and all prefixes to watch from unsynced.
+type watchKVs struct {
+	keys     [][]byte
+	vs       [][]byte
+	prefixes map[string]struct{}
+}
+
+// kvsFromUnsynced returns watchKVs. store must be mutex locked before this function.
+func kvsFromUnsynced(s *watchableStore) watchKVs {
+	// in order to find key-value pairs from unsynced watchers, we need to
+	// find min revision index, and these revisions can be used to
+	// query the backend store of key-value pairs
+	minRev := int64(math.MaxInt64)
+
+	curRev := s.store.currentRev.main
+	compactionRev := s.store.compactMainRev
+
+	prefixes := make(map[string]struct{})
+
+	for k, wm := range s.unsynced {
+		for w := range wm {
+			if w.cur > curRev {
+				panic("watcher current revision should not exceed current revision")
+			}
+			if w.cur < compactionRev {
+				// TODO: return error compacted to that watcher instead of
+				// just removing it sliently from unsynced.
+				if err := unsafeDeleteWatcher(&s.unsynced, k, w); err != nil {
+					log.Panicf("error unsafeDeleteWatcher (%v) for key %s", err, k)
+				}
+				continue
+			}
+			if minRev >= w.cur {
+				minRev = w.cur
+			}
+			if w.prefix {
+				prefixes[k] = struct{}{}
+			}
+		}
+	}
+	wv := watchKVs{prefixes: prefixes}
+
+	minBytes, maxBytes := newRevBytes(), newRevBytes()
+	revToBytes(revision{main: minRev}, minBytes)
+	revToBytes(revision{main: curRev + 1}, maxBytes)
+
+	// UnsafeRange returns keys and values. And in boltdb, keys are revisions.
+	// values are actual key-value pairs in backend.
+	tx := s.store.b.BatchTx()
+	tx.Lock()
+	wv.keys, wv.vs = tx.UnsafeRange(keyBucketName, minBytes, maxBytes, 0)
```

Comment:
we cannot do this. We have to copy the bytes returned from tx before closing the tx if you want to use it later. I would suggest to keep the old logic to use the kv in place without the copying.